scipy/sparse/linalg/_special_sparse_arrays.py

```diff
@@ -211,27 +220,36 @@ def __init__(self, grid_shape, *,
         N = np.prod(grid_shape)
         super().__init__(dtype=dtype, shape=(N, N))
 
-        indices = np.indices(grid_shape)
-        Leig = np.zeros(grid_shape)
+    def __eigenvalue_ordering(self, m):
+        grid_shape = self.grid_shape
+        if m == 0:
+            indices = np.indices(grid_shape)
+            Leig = np.zeros(grid_shape)
+        elif m > 0:
+            grid_shape_min = min(grid_shape,
+                                 tuple(np.ones_like(grid_shape) * m))
+            indices = np.indices(grid_shape_min)
+            Leig = np.zeros(grid_shape_min)
 
         for j, n in zip(indices, grid_shape):
-            if boundary_conditions == "dirichlet":
+            if self.boundary_conditions == "dirichlet":
                 Leig += -4 * np.sin(np.pi * (j + 1) / (2 * (n + 1))) ** 2
-            elif boundary_conditions == "neumann":
+            elif self.boundary_conditions == "neumann":
                 Leig += -4 * np.sin(np.pi * j / (2 * n)) ** 2
             else:  # boundary_conditions == "periodic"
                 Leig += -4 * np.sin(np.pi * np.floor((j + 1) / 2) / n) ** 2
 
-        self._eigenvalues = np.sort(Leig.ravel())
-
-    @property
-    def eigenvalues(self):
-        return self._eigenvalues
-
-    @eigenvalues.setter
-    def eigenvalues(self, value):
-        raise AttributeError('"eigenvalues" attribute is read-only and cannot '
-                             'be set.')
+        Leig_ravel = Leig.ravel()
+        ind = np.argsort(Leig_ravel)
+        _eigenvalues = Leig_ravel[ind]
+        if m > 0:
+            _eigenvalues = _eigenvalues[-m:]
+        
+        return _eigenvalues, ind
+
+    def eigenvalues(self, m=0):
```

Comment:
All of our partial eigensolvers have you specify whether you want the largest or smallest `m>0` eigenvalues with a separate argument. I would follow `lobpcg` here and add a `largest=True` argument.

```suggestion
    def eigenvalues(self, m=None, largest=True):
        """Return the requested number of eigenvalues.
        
        Parameters
        ----------
        m : int, optional
            The positive number of eigenvalues to return. If not provided,
            then all eigenvalues will be returned.
        largest : bool, optional
            Whether to return the largest ``m`` eigenvalues (default is ``True``).
            Otherwise, return the smallest ``m`` eigenvalues.
            
        Returns
        -------
        eigenvalues : float array
            The requested eigenvalues, in ascending order.
        """
```

Reply:
@rkern I currently give up on the idea to cover both cases in this PR, and only consider m >0 the largest (smallest by absolute value since they are all negative), corresponding to smooth eigenvectors normally used for testing in PDEs

Reply:
Done otherwise